python/semantic_kernel/kernel.py

```diff
@@ -106,21 +106,34 @@ def register_semantic_function(
 
         return function
 
-    async def run_async(self, *functions: Any) -> SKContext:
-        return await self.run_on_vars_async(ContextVariables(), *functions)
-
-    async def run_on_str_async(self, input_str: str, *functions: Any) -> SKContext:
-        return await self.run_on_vars_async(ContextVariables(input_str), *functions)
-
-    async def run_on_vars_async(
-        self, input_vars: ContextVariables, *functions: Any
+    async def run_async(
+            self,
+            *functions: Any,
+            input_context: Optional[SKContext] = None,
+            input_vars: Optional[ContextVariables] = None,
+            input_str: Optional[str] = None,
     ) -> SKContext:
-        context = SKContext(
-            input_vars,
-            self._memory,
-            self._skill_collection.read_only_skill_collection,
-            self._log,
-        )
+        
+        if input_context is None:
```

Comment:
I think we should handle the case where all of the 3 are not None
* copy input_str into input_vars
* copy input_vars into input_context
* use context

Reply:
right now in this case, I'm just logging warnings. Do you think it makes sense to overwrite the vars of the passed context?

Reply:
Updated to overwrite existing context if input_str or input_vars are not none